### ISSUE TITLE
Discussing breaking changes

### ISSUE THREAD

**Issue by tabatkins** [tabatkins/bikeshed#321]
Heya @mikewest @dirkschulze @annevk @rubys @fantasai @birtles, are all of you subscribed to spec-prod@w3.org?  I plan to use that mailing list to announce changes and other things to Bikeshed.  In particular, there's a big breaking change I really need to make (related to #230) that I want to give everyone heads-up on, but future changes need announcement too.

**Comment by birtles** [comment]
Signed up now.

**Comment by annevk** [comment]
I would much prefer it if you emphasized strongly on backwards compatibility. There's so many things to keep track of already.

**Comment by dirkschulze** [comment]
I agree with @annevk that it would be great if Bikeshed starts focusing on backwards compatibility now. I am signed up but won't have much time for a lot of edits the next weeks.

**Comment by mikewest** [comment]
Eh. Living standards need living tools. I've certainly spent more than a few hours cleaning up tool-induced breakage, but I suspect we could deal with that by tagging release branches so that folks who want to stay on the bleeding edge can get all the new stuff, and folks who want to stay in the past can do so easily. :)

**Comment by domenic** [comment]
Yeah, fast evolution + semantic versioning (including of the curl API) to avoid unexpected breakage would be my preferred solution

**Comment by tabatkins** [comment]
I already focus pretty strongly on backwards compat.  Things break on occasion, but unintentionally, and I fix fast.  The #230 bug I'm referring to is a long-standing footgun/issue that I'd like to fix to make Bikeshed make more sense for the future.  Being unable to provide a real title for things is annoying. :(

But back to the subject - y'all all subscribed to spec-prod?  Or would you prefer something else that's potentially less noisy and more Bikeshed-focused, like a custom mailing list?

**Comment by annevk** [comment]
Slight preference for just Bikeshed.

**Comment by astearns** [comment]
What if there was a list of bleeding-edge specs that could be used as test cases? Breaking changes are fine, but if a bikeshed change breaks anything in those specs, you also have to update the spec source to accommodate the bikeshed change? I'd much rather have everything up-to-date in my spec sources than find out about a breaking change weeks or months in the future.

**Comment by tabatkins** [comment]
Within CSSWG/FXTF, you're covered now; just delete Overview.html from the repo, and Shepherd will automatically generate for you on push (and regenerate when necessary, if link targets change). If there's problems, it'll email you (and me).

I try to run Bikeshed periodically on some other specs as full-integration tests, particularly Web Animations, as they stress a bunch of features that aren't used as much in the CSSWG.

**Comment by astearns** [comment]
I very much prefer test-before-check-in to check-in-and-see-what-breaks, so I'd suggest running more specs through your changed code more often. Perhaps even as an automated step…
But this is your project, not mine. So I'll try out letting Shepherd generate the file (after doing that myself to a temp directory before checking in, of course)

**Comment by tabatkins** [comment]
No need to generate to a temp directory.  The repo also has ignore rules for Overview.html, so you can generate it locally and just ignore it; you just have to delete it *from the repo*.

And I've finally set up a TravisCI system for auto-running my tests, so I'll get notification a few minutes after a push if I break everything.

**Comment by tabatkins** [comment]
Since I haven't yet set up the Bikeshed-only mailing list, let this suffice as a warning to everyone that I have made the title => lt change described in #230.  This is a breaking change - Bikeshed will fail to link things properly if you continue to use `title`.  For a while, Bikeshed will warn you if you having any `title` attributes on `<a>` or `<dfn>`.

**Comment by annevk** [comment]
Perhaps we can just use this issue thread for announcements? (And remove irrelevant comments now and then to make it a bit clearer.)

**Comment by tabatkins** [comment]
Sounds good. THIS IS NOW THE OFFICIAL ANNOUNCEMENT THREAD.

**Comment by tabatkins** [comment]
New (relatively minor) breaking change: @annevk convinced me to simplify the ID generation for IDL methods when there's no conflict, to generate prettier IDs.  This just means that method IDs will be of the form `dom-{interface}-{method}`, rather than `dom-{interface}-{method}{args}`, unless you specifically defined several overloads in the same spec.

**Comment by tabatkins** [comment]
Not really a breaking change, but likely to bork you at first: I just switched syntax highlighting from Prism.js to Pygments.  If you don't have the pygments package installed, Bikeshed will complain and tell you how to download it.

**Comment by tabatkins** [comment]
Not really a breaking change, but I just pushed some commits that change the format of one of the metadata files. Run `bikeshed update` or else you'll get a crash.  (I'll write the code to automatically pick up more up-to-date files from the repo copy next, I swear!)

**Comment by tabatkins** [comment]
Possibly breaking change: annevk ran into a markdown parsing bug <https://github.com/tabatkins/bikeshed/issues/623> that, if I want to fix it, requires being more strict about markup.

Specifically, I check to see if we're starting an "opaque block" (script/style/pre/xmp) by just seeing if the start tag shows up anywhere on the line.  Markdown processing happens before HTML parsing, so this fires on tags in comments or attribute values, which the HTML parser obviously won't see as a start tag.

The only way to fix this reasonably is to tighten the requirements for recognizing these blocks. Specifically, I'd change them to only be recognized when they're the first non-whitespace content on the line.

This shouldn't be a change for style/script in practice, but if you have any `<pre>` that aren't the first thing on that line (or you use the `Opaque Elements` metadata and those tags aren't the first on their line), they'll start getting markdown-processed, which probably isn't what you want.  Let me know before I change this!

**Comment by tabatkins** [comment]
Never mind, this wasn't necessary. Fixed the problem more directly by just stripping comments at Markdown-parsing time.

This is not a breaking change, because (a) the minify setting defaults to removing comments, and (b) regardless of the minify setting, the serializer was accidentally just skipping comments anyway, so your output has been comment-less without any recourse for some time now.

I'll delete this comment and the preceding in a sec. Just wanted to give everyone an update that the previous comment can be ignored anyway.

**Comment by tabatkins** [comment]
Just pushed some internal rearchitecting of how I handle refs, separating out local/anchor-block/foreign into separate data structures. Shouldn't be any visible change (I ran it on a couple of specs to make sure), but if you see anything confusing, let me know.

On the plus side, this is a minor but significant perf win; DOM processes about 10% faster!

**Comment by tabatkins** [comment]
And... Bikeshed is now a Python 3 project. It'll now print out a nice refusal message if you try to run it on Py2.

Install instructions forthcoming, but the most relevant bit is that it requires Python 3.7; this isn't the default py3 install on many OSes, so I recommend using pyenv to install it. [Here's a tutorial.](https://amaral.northwestern.edu/resources/guides/pyenv-tutorial)

**Comment by tabatkins** [comment]
@frivoal just pushed a very nice set of commits improving various little bits of the `<wpt>` element, most notably the new `WPT Display: closed` value and associated styles. When this is turned on, the `<wpt>` element is a `<details>` that starts closed, with the summary being a small, out-of-the-way little "TESTS" that doesn't visually interrupt at all.

This is looking so good that I think I'll flip it to the default soon, for non-TR documents. If you're currently use `<wpt>` and definitely do *not* want tests to show in your document, add a `WPT Display: none` metadata to your spec or your group's `defaults.include` file.

We flipped it on by default in the CSSWG just a few minutes ago, so in a little bit (once everything regens) you'll be able to see how it looks on any CSSWG spec - I recommend checking [Color 4](https://drafts.csswg.org/css-color-4/) and searching for "tests".

**Comment by svgeesus** [comment]
> This is looking so good that I think I'll flip it to the default soon, for non-TR documents. 

I would prefer to see them like this on /TR documents, too.

**Comment by tabatkins** [comment]
Ah yeah it's been nearly a year and the experience has been great, I'll just flip the default

**Comment by tabatkins** [comment]
As part of my quest to add typing hints to all of Bikeshed, I'm also ending up refreshing some old confusing code that dates back to my earliest implementation. Most recently, I did a decent rewrite of my biblio handling, which will likely cause some changes - in particular, I believe I now *correctly* handle the `Default Biblio Status` metadata (and its default, which is based on your spec's `Status`) in all places, which from the testsuite diff will probably affect a lot of references, switching them from snapshots to EDs or vice versa.

If anything looks *wrong*, please let me know.

**Comment by frivoal** [comment]
Notice of breaking change, effective for a while already and documented in https://github.com/speced/bikeshed/issues/1773#issuecomment-1599680734:
The ABSTRACTATTR macro is no longer supported by bikeshed. If you use that in your templates or documents, they need updating.

**Comment by tabatkins** [comment]
And another breaking change, because I just discovered today that Python 3.9 (the current minimum supported version) hit EOL back in October, so it's no longer receiving security updates. Doing some looking around, it seems like 3.12 is a reasonable version to bump to, so I've now published Bikeshed v6.1 (which requires 3.12 in practice and prints an error if you're on a lower version, but doesn't tell `pip` that 3.12 is required yet) and Bikeshed v7.0 (which tells pip that 3.12 is needed).

Hopefully, doing it this way will avoid the problem I had *last* python version bump, where people with old CI OS choices that had an unsupported Python version would just get stuck on the old Bikeshed. Now, they'll instead grab version 6.1, which technically works on 3.9 but dies on startup with a useful error message. Anyone with the necessary new Python version will get v7 which works.
